File: backend/app/api/v1/payments.py
"""Payment and credit endpoints for Lextract.

Provides checkout session creation, credit usage, balance queries,
and payment history. Authenticated users always get the full flow;
anonymous guests can pay with just an email (guest checkout).
"""

import logging
from datetime import UTC, datetime

from fastapi import APIRouter, HTTPException, Query, status

from app.core.dependencies import CurrentUser, OptionalUser
from app.core.exceptions import (
    ConflictError,
    InsufficientCreditsError,
    NotFoundError,
    StripeError,
)
from app.database.client import NeonClientManager
from app.models.user import AnonymousSession, User
from app.schemas.credit import (
    CreditBalanceResponse,
    CreditTransactionResponse,
    PaymentHistoryResponse,
    PaymentRecord,
    UseCreditRequest,
    UseCreditResponse,
)
from app.schemas.payment import CheckoutRequest, CheckoutResponse
from app.services.credit_service import get_credit_service
from app.services.stripe_service import get_stripe_service

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/payments", tags=["Payments"])


@router.post(
    "/checkout",
    response_model=CheckoutResponse,
    status_code=status.HTTP_200_OK,
    summary="Create a Stripe Checkout Session",
)
async def create_checkout(
    body: CheckoutRequest,
    caller: OptionalUser,
) -> CheckoutResponse:
    """Create a Stripe Checkout Session for a one-time purchase.

    Accepts both authenticated users and anonymous guests.  For guests,
    ``guest_email`` must be provided; the email is stored on the extraction row
    and forwarded to Stripe so the receipt is delivered automatically.

    Returns the Stripe Checkout URL and session ID for client-side redirect.
    """
    # Determine who is checking out
    is_authenticated_user = isinstance(caller, User)
    is_anonymous_session = isinstance(caller, AnonymousSession)
    is_guest = is_anonymous_session and body.guest_email is not None

    if not is_authenticated_user and not is_guest:
        # Unauthenticated and no guest_email — require auth
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Authentication required or guest_email must be provided",
            headers={"WWW-Authenticate": "Bearer"},
        )

    stripe_svc = get_stripe_service()

    # For authenticated users, use their user_id.
    # For guests, user_id is empty — the webhook will create the account.
    # Use isinstance narrowing so mypy can resolve .id safely.
    user_id = str(caller.id) if isinstance(caller, User) else ""
    guest_email = body.guest_email if is_guest else None
    anonymous_session_id = (
        str(caller.id) if isinstance(caller, AnonymousSession) and is_guest else None
    )

    if body.product_type == "single" and not body.extraction_id:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="extraction_id is required for single purchases",
        )

    # Validate extraction ownership before placing the ID in Stripe metadata.
    if body.product_type == "single" and body.extraction_id:
        db = NeonClientManager.get_service_client()
        now = datetime.now(UTC).isoformat()
        if is_authenticated_user:
            owned = (
                db.table("extractions")
                .select("id")
                .eq("id", body.extraction_id)
                .eq("user_id", user_id)
                .eq("payment_status", "unpaid")
                .maybe_single()
                .execute()
            )
            if not owned.data:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail="Extraction not found",
                )
        elif is_guest:
            update_result = (
                db.table("extractions")
                .update({"guest_email": guest_email, "updated_at": now})
                .eq("id", body.extraction_id)
                .eq("anonymous_session_id", anonymous_session_id)
                .is_("user_id", "null")
                .eq("payment_status", "unpaid")
                .execute()
            )
            if not update_result.data:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail="Extraction not found",
                )

    try:
        session = stripe_svc.create_checkout_session(
            user_id=user_id,
            product_type=body.product_type,
            success_url=body.success_url,
            cancel_url=body.cancel_url,
            extraction_id=body.extraction_id,
            guest_email=guest_email,
            anonymous_session_id=anonymous_session_id,
        )
    except StripeError as exc:
        logger.error("Stripe checkout creation failed: %s", exc.message)
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY,
            detail="Payment provider error — please try again",
        ) from exc

    return CheckoutResponse(
        checkout_url=str(session.url),
        session_id=str(session.id),
    )


@router.post(
    "/use-credit",
    response_model=UseCreditResponse,
    status_code=status.HTTP_200_OK,
    summary="Use one credit to unlock an extraction",
)
async def use_credit(
    body: UseCreditRequest,
    user: CurrentUser,
) -> UseCreditResponse:
    """Deduct one credit and mark the extraction as paid.

    Returns 402 if the user has insufficient credits.
    Returns 409 if a concurrent modification is detected.
    """
    credit_svc = get_credit_service()

    try:
        result = await credit_svc.use_credit(
            user_id=str(user.id),
            extraction_id=body.extraction_id,
        )
    except InsufficientCreditsError as exc:
        raise HTTPException(
            status_code=status.HTTP_402_PAYMENT_REQUIRED,
            detail=str(exc),
        ) from exc
    except NotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc
    except ConflictError as exc:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=str(exc),
        ) from exc

    return UseCreditResponse(
        success=True,
        new_balance=result["new_balance"],
        extraction_id=result["extraction_id"],
    )


@router.get(
    "/credits",
    response_model=CreditBalanceResponse,
    status_code=status.HTTP_200_OK,
    summary="Get current credit balance and recent transactions",
)
async def get_credits(
    user: CurrentUser,
) -> CreditBalanceResponse:
    """Return the user's credit balance and recent transactions."""
    credit_svc = get_credit_service()

    balance = credit_svc.get_balance(str(user.id))
    transactions = credit_svc.get_recent_transactions(str(user.id))

    return CreditBalanceResponse(
        balance=balance,
        recent_transactions=[
            CreditTransactionResponse(
                id=str(t["id"]),
                amount=t["amount"],
                balance_after=t["balance_after"],
                description=t["description"],
                created_at=t["created_at"],
            )
            for t in transactions
        ],
    )


@router.get(
    "/history",
    response_model=PaymentHistoryResponse,
    status_code=status.HTTP_200_OK,
    summary="Get paginated payment history",
    responses={
        200: {
            "description": "Paginated list of completed payments for the user.",
            "content": {
                "application/json": {
                    "example": {
                        "payments": [
                            {
                                "id": "11111111-1111-4111-a111-111111111111",
                                "payment_type": "single",
                                "amount_cents": 1500,
                                "currency": "usd",
                                "status": "completed",
                                "created_at": "2026-05-01T12:00:00Z",
                            },
                            {
                                "id": "22222222-2222-4222-a222-222222222222",
                                "payment_type": "credit_pack_5",
                                "amount_cents": 6000,
                                "currency": "usd",
                                "status": "completed",
                                "created_at": "2026-04-20T09:30:00Z",
                            },
                        ],
                        "total": 2,
                        "page": 1,
                        "page_size": 20,
                    }
                }
            },
        },
        401: {"description": "Authentication required."},
        422: {"description": "Invalid pagination parameters."},
    },
)
async def get_payment_history(
    user: CurrentUser,
    page: int = Query(default=1, ge=1, description="Page number (1-indexed)"),
    page_size: int = Query(default=20, ge=1, le=100, description="Items per page"),
) -> PaymentHistoryResponse:
    """Return paginated payment history for the authenticated user.

    Each row reflects an immutable Stripe-backed payment. Credit-pack
    purchases additionally appear as positive entries in the credit
    ledger (`GET /payments/credits`); single-extraction purchases do not
    grant credits and therefore do not produce ledger rows.
    """
    credit_svc = get_credit_service()

    payments, total = credit_svc.get_payment_history(
        user_id=str(user.id),
        page=page,
        page_size=page_size,
    )

    return PaymentHistoryResponse(
        payments=[
            PaymentRecord(
                id=str(p["id"]),
                payment_type=p["payment_type"],
                amount_cents=p["amount_cents"],
                currency=p.get("currency", "usd"),
                status=p["status"],
                created_at=p["created_at"],
            )
            for p in payments
        ],
        total=total,
        page=page,
        page_size=page_size,
    )

File: workers/api/src/routes/payments.ts
import { Hono } from 'hono'
import { z } from 'zod'
import type { Context } from 'hono'

import {
  CheckoutNotFoundError,
  CheckoutValidationError,
  ConflictPaymentError,
  InsufficientCreditsPaymentError,
  NotFoundPaymentError,
  getCredits,
  getPaymentHistory,
  useCredit,
  validateCheckoutOwnership,
} from '../repositories/payments'
import type {
  CreditsSummary,
  PaymentHistory,
  PaymentHistoryInput,
  UseCreditInput,
  UseCreditResult,
} from '../repositories/payments'
import { createRequireUserAuth, createAuthMiddleware } from '../middleware/auth'
import type { AuthContext, AuthDependencies } from '../services/neon-auth'
import { createStripeCheckoutSession } from '../services/stripe'
import type { ProductType, StripeCheckoutInput, StripeCheckoutResult } from '../services/stripe'
import { StripeProviderError } from '../services/stripe'
import type { AppBindings, Env } from '../types'

const productTypeSchema = z.enum(['single', 'credit_pack_5', 'credit_pack_10'])
const httpsOrLocalUrl = z.string().min(1).max(2048).transform((value, context) => {
  const trimmed = value.trim()
  let parsed: URL
  try {
    parsed = new URL(trimmed)
  } catch {
    context.addIssue({ code: 'custom', message: 'URL must be valid' })
    return z.NEVER
  }
  if (
    parsed.protocol === 'https:' ||
    (parsed.protocol === 'http:' &&
      (parsed.hostname === 'localhost' || parsed.hostname === '127.0.0.1'))
  ) {
    return trimmed
  }
  context.addIssue({
    code: 'custom',
    message: 'URL must use https:// or localhost http://',
  })
  return z.NEVER
})
const optionalEmail = z
  .string()
  .trim()
  .email()
  .optional()
  .nullable()
  .transform((value) => (value === null ? undefined : value))
const checkoutSchema = z.object({
  cancel_url: httpsOrLocalUrl,
  extraction_id: z.string().min(1).optional().nullable(),
  guest_email: optionalEmail,
  product_type: productTypeSchema,
  success_url: httpsOrLocalUrl,
})
const useCreditSchema = z.object({
  extraction_id: z.string().min(1),
})

export interface CheckoutSessionInput {
  userId?: string
  productType: ProductType
  successUrl: string
  cancelUrl: string
  extractionId?: string
  guestEmail?: string
  anonymousSessionId?: string
}

export interface PaymentRouteDependencies {
  authDependencies?: AuthDependencies
  createCheckoutSession(input: CheckoutSessionInput, env: Env): Promise<StripeCheckoutResult>
  useCredit(input: UseCreditInput, env: Env): Promise<UseCreditResult>
  getCredits(userId: string, env: Env): Promise<CreditsSummary>
  getPaymentHistory(input: PaymentHistoryInput, env: Env): Promise<PaymentHistory>
}

type UserAuthContext = Extract<AuthContext, { kind: 'user' }>

function currentUser(c: Context<AppBindings>): UserAuthContext {
  // Safe because every caller is behind createRequireUserAuth middleware.
  return c.get('authContext') as UserAuthContext
}

async function defaultCreateCheckoutSession(
  input: CheckoutSessionInput,
  env: Env,
): Promise<StripeCheckoutResult> {
  await validateCheckoutOwnership(input, env)
  const stripeInput: StripeCheckoutInput = {
    cancelUrl: input.cancelUrl,
    productType: input.productType,
    successUrl: input.successUrl,
  }
  if (input.userId !== undefined) {
    stripeInput.userId = input.userId
  }
  if (input.extractionId !== undefined) {
    stripeInput.extractionId = input.extractionId
  }
  if (input.guestEmail !== undefined) {
    stripeInput.guestEmail = input.guestEmail
  }
  if (input.anonymousSessionId !== undefined) {
    stripeInput.anonymousSessionId = input.anonymousSessionId
  }
  return createStripeCheckoutSession(stripeInput, env)
}

export function defaultPaymentRouteDependencies(): PaymentRouteDependencies {
  return {
    createCheckoutSession: defaultCreateCheckoutSession,
    getCredits,
    getPaymentHistory,
    useCredit,
  }
}

function creditsResponse(summary: CreditsSummary): Record<string, unknown> {
  return {
    balance: summary.balance,
    recent_transactions: summary.recentTransactions.map((transaction) => ({
      amount: transaction.amount,
      balance_after: transaction.balanceAfter,
      created_at: transaction.createdAt,
      description: transaction.description,
      id: transaction.id,
    })),
  }
}

function historyResponse(history: PaymentHistory): Record<string, unknown> {
  return {
    page: history.page,
    page_size: history.pageSize,
    payments: history.payments.map((payment) => ({
      amount_cents: payment.amountCents,
      created_at: payment.createdAt,
      currency: payment.currency,
      id: payment.id,
      payment_type: payment.paymentType,
      status: payment.status,
    })),
    total: history.total,
  }
}

function mapPaymentError(error: unknown): Response | null {
  if (error instanceof CheckoutValidationError) {
    return Response.json({ detail: error.message }, { status: 422 })
  }
  if (error instanceof CheckoutNotFoundError || error instanceof NotFoundPaymentError) {
    return Response.json({ detail: error.message }, { status: 404 })
  }
  if (error instanceof ConflictPaymentError) {
    return Response.json({ detail: error.message }, { status: 409 })
  }
  if (error instanceof InsufficientCreditsPaymentError) {
    return Response.json({ detail: error.message }, { status: 402 })
  }
  if (error instanceof StripeProviderError) {
    return Response.json({ detail: error.message }, { status: 502 })
  }
  return null
}

export function createPaymentsRoutes(
  dependencies: PaymentRouteDependencies = defaultPaymentRouteDependencies(),
): Hono<AppBindings> {
  const payments = new Hono<AppBindings>()
  const auth = createAuthMiddleware(dependencies.authDependencies)
  const requireUser = createRequireUserAuth(dependencies.authDependencies)

  payments.post('/checkout', auth, async (c) => {
    const parsed = checkoutSchema.safeParse(await c.req.json())
    if (!parsed.success) {
      return c.json({ detail: 'Invalid checkout request' }, 422)
    }
    const body = parsed.data
    const authContext = c.get('authContext')
    const isUser = authContext.kind === 'user'
    const isGuest =
      authContext.kind === 'anonymous' && body.guest_email !== undefined
    if (!isUser && !isGuest) {
      return c.json(
        { detail: 'Authentication required or guest_email must be provided' },
        401,
      )
    }
    if (body.product_type === 'single' && !body.extraction_id) {
      return c.json(
        { detail: 'extraction_id is required for single purchases' },
        422,
      )
    }

    try {
      const checkoutInput: CheckoutSessionInput = {
        cancelUrl: body.cancel_url,
        productType: body.product_type,
        successUrl: body.success_url,
      }
      if (authContext.kind === 'user') {
        checkoutInput.userId = authContext.id
      }
      if (authContext.kind === 'anonymous') {
        checkoutInput.anonymousSessionId = authContext.id
      }
      if (body.extraction_id !== undefined && body.extraction_id !== null) {
        checkoutInput.extractionId = body.extraction_id
      }
      if (body.guest_email !== undefined) {
        checkoutInput.guestEmail = body.guest_email
      }
      const session = await dependencies.createCheckoutSession(
        checkoutInput,
        c.env,
      )
      return c.json({
        checkout_url: session.checkoutUrl,
        session_id: session.sessionId,
      })
    } catch (error) {
      const response = mapPaymentError(error)
      if (response !== null) {
        return response
      }
      throw error
    }
  })

  payments.post('/use-credit', requireUser, async (c) => {
    const authContext = currentUser(c)
    const parsed = useCreditSchema.safeParse(await c.req.json())
    if (!parsed.success) {
      return c.json({ detail: 'Invalid credit request' }, 422)
    }
    const body = parsed.data
    try {
      const result = await dependencies.useCredit(
        { extractionId: body.extraction_id, userId: authContext.id },
        c.env,
      )
      return c.json({
        extraction_id: result.extractionId,
        new_balance: result.newBalance,
        success: true,
      })
    } catch (error) {
      const response = mapPaymentError(error)
      if (response !== null) {
        return response
      }
      throw error
    }
  })

  payments.get('/credits', requireUser, async (c) => {
    const authContext = currentUser(c)
    return c.json(creditsResponse(await dependencies.getCredits(authContext.id, c.env)))
  })

  payments.get('/history', requireUser, async (c) => {
    const authContext = currentUser(c)
    const page = Number(c.req.query('page') ?? '1')
    const pageSize = Number(c.req.query('page_size') ?? '20')
    if (
      !Number.isInteger(page) ||
      !Number.isInteger(pageSize) ||
      page < 1 ||
      pageSize < 1 ||
      pageSize > 100
    ) {
      return c.json({ detail: 'Invalid pagination parameters' }, 422)
    }
    return c.json(
      historyResponse(
        await dependencies.getPaymentHistory(
          { page, pageSize, userId: authContext.id },
          c.env,
        ),
      ),
    )
  })

  return payments
}

export const paymentsRoutes = createPaymentsRoutes()
